File: commons-core/src/main/java/xyz/wagyourtail/commons/position/Pos3.java
package xyz.wagyourtail.commons.position;

import java.util.Objects;

public class Pos3 extends Pos2 {
    public static final Pos3 ZERO = new Pos3(0, 0, 0);

    public final int z;

    public Pos3(int x, int y, int z) {
        super(x, y);
        this.z = z;
    }

    public static Pos3 max(Pos3 p1, Pos3 p2) {
        return new Pos3(Math.max(p1.x, p2.x), Math.max(p1.y, p2.y), Math.max(p1.z, p2.z));
    }

    public static Pos3 min(Pos3 p1, Pos3 p2) {
        return new Pos3(Math.max(p1.x, p2.x), Math.min(p1.y, p2.y), Math.min(p1.z, p2.z));
    }

    public static Pos3 abs(Pos3 p) {
        return new Pos3(Math.abs(p.x), Math.abs(p.y), Math.abs(p.z));
    }

    public Pos3 up() {
        return new Pos3(x, y - 1, z);
    }

    public Pos3 down() {
        return new Pos3(x, y + 1, z);
    }

    public Pos3 left() {
        return new Pos3(x - 1, y, z);
    }

    public Pos3 right() {
        return new Pos3(x + 1, y, z);
    }

    public Pos3 forward() {
        return new Pos3(x, y, z + 1);
    }

    public Pos3 back() {
        return new Pos3(x, y, z - 1);
    }

    @Override
    public Pos3 inverse() {
        return new Pos3(-this.x, -this.y, -this.z);
    }

    public Pos3 plus(Pos3 other) {
        return new Pos3(this.x + other.x, this.y + other.y, this.z + other.z);
    }

    public Pos3 plus(int x, int y, int z) {
        return new Pos3(this.x + x, this.y + y, this.z + z);
    }

    @Override
    public Pos3 plus(int scalar) {
        return new Pos3(this.x + scalar, this.y + scalar, this.z + scalar);
    }

    public Pos3 minus(Pos3 other) {
        return new Pos3(this.x - other.x, this.y - other.y, this.z - other.z);
    }

    public Pos3 minus(int x, int y, int z) {
        return new Pos3(this.x - x, this.y - y, this.z - z);
    }

    @Override
    public Pos3 minus(int scalar) {
        return new Pos3(this.x - scalar, this.y - scalar, this.z - scalar);
    }

    public Pos3 times(Pos3 other) {
        return new Pos3(this.x * other.x, this.y * other.y, this.z * other.z);
    }

    public Pos3 times(int x, int y, int z) {
        return new Pos3(this.x * x, this.y * y, this.z * z);
    }

    @Override
    public Pos3 times(int scalar) {
        return new Pos3(this.x * scalar, this.y * scalar, this.z * scalar);
    }

    public Pos3 div(Pos3 other) {
        return new Pos3(this.x / other.x, this.y / other.y, this.z / other.z);
    }

    public Pos3 div(int x, int y, int z) {
        return new Pos3(this.x / x, this.y / y, this.z / z);
    }

    @Override
    public Pos3 div(int scalar) {
        return new Pos3(this.x / scalar, this.y / scalar, this.z / scalar);
    }

    public Pos3 rem(Pos3 other) {
        return new Pos3(x % other.x, y % other.y, z % other.z);
    }

    public Pos3 rem(int x, int y, int z) {
        return new Pos3(this.x % x, this.y % y, this.z % z);
    }

    @Override
    public Pos3 rem(int scalar) {
        return new Pos3(x % scalar, y % scalar, z % scalar);
    }

    public Pos3 mod(Pos3 other) {
        int x = this.x % other.x;
        int y = this.y % other.y;
        int z = this.z % other.z;
        if (x < 0) x += other.x;
        if (y < 0) y += other.y;
        if (z < 0) z += other.z;
        return new Pos3(x, y, z);
    }

    public Pos3 mod(int x, int y, int z) {
        int x2 = this.x % x;
        int y2 = this.y % y;
        int z2 = this.z % z;
        if (x2 < 0) x2 += x;
        if (y2 < 0) y2 += y;
        if (z2 < 0) z2 += z;
        return new Pos3(x2, y2, z2);
    }

    @Override
    public Pos3 mod(int scalar) {
        int x = this.x % scalar;
        int y = this.y % scalar;
        int z = this.z % scalar;
        if (x < 0) x += scalar;
        if (y < 0) y += scalar;
        if (z < 0) z += scalar;
        return new Pos3(x, y, z);
    }

    @Override
    public Pos3D toDouble() {
        return new Pos3D(x, y, z);
    }

    @Override
    public double magnitude() {
        return Math.sqrt(x * x + y * y + z * z);
    }

    @Override
    public int magnitudeSquared() {
        return x * x + y * y + z * z;
    }

    public double dot(Pos3 other) {
        return super.dot(other) + z * other.z;
    }

    public Pos3 cross(Pos3 other) {
        return new Pos3(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z, this.x * other.y - this.y * other.x);
    }

    @Override
    public String toString() {
        return "(" + x + ", " + y + ", " + z + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Pos3)) return false;
        if (!super.equals(o)) return false;
        Pos3 pos3 = (Pos3) o;
        return z == pos3.z;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), z);
    }

}

File: commons-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/position/Pos3.kt
package xyz.wagyourtail.commonskt.position

import kotlin.math.abs
import kotlin.math.sqrt

class Pos3(x: Int, y: Int, val z: Int) : Pos2(x, y) {
    companion object {
        val ZERO = Pos3(0, 0, 0)
    }

    operator fun component3() = z

    override val up: Pos3
        get() = Pos3(this.x, this.y - 1, this.z)

    override val down: Pos3
        get() = Pos3(this.x, this.y + 1, this.z)

    override val left: Pos3
        get() = Pos3(this.x - 1, this.y, this.z)

    override val right: Pos3
        get() = Pos3(this.x + 1, this.y, this.z)

    val forward
        get() = Pos3(this.x, this.y, this.z + 1)

    val back
        get() = Pos3(this.x, this.y, this.z - 1)

    operator fun rangeTo(pos: Pos3): Sequence<Pos3> {
        return sequence {
            for (z in z..pos.z) {
                for (y in y..pos.y) {
                    for (x in x..pos.x) {
                        yield(Pos3(x, y, z))
                    }
                }
            }
        }
    }

    operator fun rangeUntil(pos: Pos3): Sequence<Pos3> {
        return sequence {
            for (z in z..<pos.z) {
                for (y in y..<pos.y) {
                    for (x in x + 1..<x) {
                        yield(Pos3(x, y, z))
                    }
                }
            }
        }
    }

    override operator fun unaryMinus(): Pos3 {
        return Pos3(-this.x, -this.y, -this.z)
    }

    operator fun plus(other: Pos3): Pos3 {
        return Pos3(this.x + other.x, this.y + other.y, this.z + other.z)
    }

    fun plus(x: Int, y: Int, z: Int): Pos3 {
        return Pos3(this.x + x, this.y + y, this.z + z)
    }

    override operator fun plus(scalar: Int): Pos3 {
        return Pos3(this.x + scalar, this.y + scalar, this.z + scalar)
    }

    fun minus(other: Pos3): Pos3 {
        return Pos3(this.x - other.x, this.y - other.y, this.z - other.z)
    }

    fun minus(x: Int, y: Int, z: Int): Pos3 {
        return Pos3(this.x - x, this.y - y, this.z - z)
    }

    override operator fun minus(scalar: Int): Pos3 {
        return Pos3(this.x - scalar, this.y - scalar, this.z - scalar)
    }

    operator fun times(other: Pos3): Pos3 {
        return Pos3(this.x * other.x, this.y * other.y, this.z * other.z)
    }

    fun times(x: Int, y: Int, z: Int): Pos3 {
        return Pos3(this.x * x, this.y * y, this.z * z)
    }

    override operator fun times(scalar: Int): Pos3 {
        return Pos3(this.x * scalar, this.y * scalar, this.z * scalar)
    }

    operator fun div(other: Pos3): Pos3 {
        return Pos3(this.x / other.x, this.y / other.y, this.z / other.z)
    }

    fun div(x: Int, y: Int, z: Int): Pos3 {
        return Pos3(this.x / x, this.y / y, this.z / z)
    }

    override operator fun div(scalar: Int): Pos3 {
        return Pos3(this.x / scalar, this.y / scalar, this.z / scalar)
    }

    operator fun rem(other: Pos3): Pos3 {
        return Pos3(x % other.x, y % other.y, z % other.z)
    }

    fun rem(x: Int, y: Int, z: Int): Pos3 {
        return Pos3(this.x % x, this.y % y, this.z % z)
    }

    override operator fun rem(scalar: Int): Pos3 {
        return Pos3(x % scalar, y % scalar, z % scalar)
    }

    infix fun mod(other: Pos3): Pos3 {
        var x = this.x % other.x
        var y = this.y % other.y
        var z = this.z % other.z
        if (x < 0) x += other.x
        if (y < 0) y += other.y
        if (z < 0) z += other.z
        return Pos3(x, y, z)
    }

    fun mod(x: Int, y: Int, z: Int): Pos3 {
        var x2 = this.x % x
        var y2 = this.y % y
        var z2 = this.z % z
        if (x2 < 0) x2 += x
        if (y2 < 0) y2 += y
        if (z2 < 0) z2 += z
        return Pos3(x2, y2, z2)
    }

    override infix fun mod(scalar: Int): Pos3 {
        var x = this.x % scalar
        var y = this.y % scalar
        var z = this.z % scalar
        if (x < 0) x += scalar
        if (y < 0) y += scalar
        if (z < 0) z += scalar
        return Pos3(x, y, z)
    }

    override fun toDouble(): Pos3D {
        return Pos3D(x.toDouble(), y.toDouble(), z.toDouble())
    }

    fun distanceTo(other: Pos3): Double {
        return sqrt(distanceToSquared(other).toDouble())
    }

    fun distanceToSquared(other: Pos3): Int {
        return (x - other.x) * (x - other.x) + (y - other.y) * (y - other.y) + (z - other.z) * (z - other.z)
    }

    fun manhattanDistanceTo(other: Pos3): Int {
        return abs(x - other.x) + abs(y - other.y) + abs(z - other.z)
    }

    override val magnitude: Double
        get() {
            return sqrt((x * x + y * y + z * z).toDouble())
        }

    override val magnitudeSquared: Int
        get() {
            return x * x + y * y + z * z
        }

    override val manhattanMagnitude: Int
        get() = abs(x) + abs(y) + abs(z)

    fun dot(other: Pos3): Double {
        return (super.dot(other) + z * other.z).toDouble()
    }

    fun cross(other: Pos3): Pos3 {
        return Pos3(
            this.y * other.z - this.z * other.y,
            this.z * other.x - this.x * other.z,
            this.x * other.y - this.y * other.x
        )
    }

    override fun toString(): String {
        return "($x, $y, $z)"
    }

    override fun equals(other: Any?): Boolean {
        return super.equals(other) && (other is Pos3 && z == other.z)
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + z.hashCode()
        return result
    }
}
